torch_xla/distributed/xla_backend.py

```diff
@@ -148,6 +148,9 @@ def send(self, tensors, dst_rank, tag=0):
     for t in tensors:
       channel_id = self.make_send_channel_id(dst_rank, tag)
       token = xm.send(t, channel_id)
+      # Make the sent tensor depend on the token, such that the `send`
+      # op can actually be built into the computation graph.
+      t.data.data = token
```

Comment:
so the assumption here is that this process will not use the `tensors` after it is being sent? Is this also specified in the pytorch ddp?

I find it a bit confusing that after the `send` the tensor becomes a token. If you want to force the dependency another thing you could try is use `xla::Tuple` and then untuple to bind the `token` and `input` (like what I do in https://github.com/pytorch/xla/pull/3493/files#diff-c561cfa200169af27f0f5ff1db5e5449e1398fce312de49e0a4c867efeea55bdR38). This will also force `send` and `token` show up in the HLO as long as one of the two is used in later graph.

Reply:
for example you can do something like
```
BuildSendWithToken(){
  xla::XlaOp result_token = xla::SendWithToken(input, token, channel_handle);

    xla::XlaOp tuple_res = xla::Tuple(inputs[0].builder(), {result_token, input});
    xla::XlaOp new_input = xla::GetTupleElement(tuple_res, 1)

  return {new_input, result_token};
}
```

then I think you can do 
```
t.data.data = new_input
```